src/_pytest/timing.py

```diff
@@ -20,6 +21,47 @@
     from pytest import MonkeyPatch
 
 
+@dataclasses.dataclass(frozen=True)
+class Instant:
+    """
+    Represents an instant in time, used to both get the timestamp value and to measure
+    the duration of a time span.
+
+    Inspired by Rust's `std::time::Instant`.
+    """
+
+    # Creation time of this instant, using time.time(), to measure actual time.
+    # Use a `lambda` to initialize the default to correctly get the mocked time via `MockTiming`.
+    time: float = dataclasses.field(default_factory=lambda: time(), init=False)
+
+    # Performance counter tick of the instant, used to measure precise elapsed time.
+    # Use a `lambda` to initialize the default to correctly get the mocked time via `MockTiming`.
+    perf_count: float = dataclasses.field(
+        default_factory=lambda: perf_counter(), init=False
+    )
+
+    def duration(self) -> Duration:
+        """Measure the duration since `Instant` was created."""
+        return Duration(start=self, stop=Instant())
+
+    def as_utc(self) -> datetime:
+        """Instant as UTC datetime."""
+        return datetime.fromtimestamp(self.time, timezone.utc)
+
+
+@dataclasses.dataclass(frozen=True)
+class Duration:
+    """A span of time as measured by `Instant.duration()`."""
+
+    start: Instant
+    stop: Instant
```

Comment:
im tempted to suggest
```suggestion
    stop: Instant = dataclasses.field(default_factory=Instant)
```

but it may be best left to the explicit method in the instant

Reply:
Yeah, I do not expect `Duration` to be manually instantiated.